File: DataEngineering/03_Data_Ingestion/projects/python/02_web_scraper_to_db/main.py
import os
import logging
import requests
from bs4 import BeautifulSoup
from sqlalchemy import create_engine, Column, Integer, String
from sqlalchemy.orm import declarative_base, sessionmaker
from requests.adapters import HTTPAdapter
from urllib3.util.retry import Retry

logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')

Base = declarative_base()

class ArticleTitle(Base):
    __tablename__ = 'article_titles'
    id = Column(Integer, primary_key=True, autoincrement=True)
    title = Column(String, nullable=False)

def get_session() -> requests.Session:
    session = requests.Session()
    retry = Retry(total=3, backoff_factor=1, status_forcelist=[500, 502, 503, 504])
    adapter = HTTPAdapter(max_retries=retry)
    session.mount("http://", adapter)
    session.mount("https://", adapter)
    return session

def init_db(db_url: str):
    engine = create_engine(db_url)
    Base.metadata.create_all(engine)
    return sessionmaker(bind=engine)

def fetch_titles(url: str) -> list[str]:
    logging.info(f"Fetching titles from {url}...")
    session = get_session()
    response = session.get(url, timeout=15)
    response.raise_for_status()
    
    soup = BeautifulSoup(response.text, 'html.parser')
    # Assuming scraping a typical tech blog with h2 tags for article titles
    titles = [h2.get_text(strip=True) for h2 in soup.find_all('h2')]
    return titles

def save_titles(db_session, titles: list[str]):
    session = db_session()
    try:
        for t in titles:
            if t:
                record = ArticleTitle(title=t)
                session.add(record)
        session.commit()
        logging.info("Titles successfully saved to database.")
    except Exception as e:
        session.rollback()
        logging.error(f"Failed to save titles: {e}")
        raise
    finally:
        session.close()

def main():
    logging.info('Starting Data Engineering Project: 02_web_scraper_to_db')
    target_url = os.getenv("SCRAPE_URL", "https://news.ycombinator.com/") 
    # Just an example target, structure may vary. 
    db_url = os.getenv("DATABASE_URL", "postgresql://user:password@localhost:5432/scraper_db")
    
    try:
        db_session_factory = init_db(db_url)
        titles = fetch_titles(target_url)
        logging.info(f"Extracted {len(titles)} titles. Saving...")
        save_titles(db_session_factory, titles)
    except Exception as e:
        logging.error("Scraper pipeline failed.", exc_info=True)

if __name__ == '__main__':
    main()

File: DataEngineering/03_Data_Ingestion/projects/golang/02_web_scraper_to_db/main.go
package main

import (
	"context"
	"database/sql"
	"io"
	"log"
	"net/http"
	"regexp"
	"time"

	_ "github.com/lib/pq"
)

func fetchHTML(ctx context.Context, url string) (string, error) {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		return "", err
	}

	client := &http.Client{Timeout: 10 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", err
	}

	return string(body), nil
}

func initDB(dbURL string) *sql.DB {
	db, err := sql.Open("postgres", dbURL)
	if err != nil {
		log.Fatalf("Failed to open DB: %v", err)
	}

	sqlStmt := `
	CREATE TABLE IF NOT EXISTS article_titles (
		id SERIAL PRIMARY KEY,
		title TEXT NOT NULL,
		created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
	);
	`
	if _, err := db.Exec(sqlStmt); err != nil {
		log.Fatalf("Failed to create table: %v", err)
	}
	return db
}

func main() {
	log.Println("Starting Data Engineering Project: 02_web_scraper_to_db (Real Target & Postgres)")

	targetURL := os.Getenv("TARGET_URL")
	if targetURL == "" {
		targetURL = "https://news.ycombinator.com/"
	}

	dbURL := os.Getenv("DATABASE_URL")
	if dbURL == "" {
		dbURL = "postgres://postgres:password@localhost:5432/scraper_db?sslmode=disable"
	}

	db := initDB(dbURL)
	defer db.Close()

	ctx, cancel := context.WithTimeout(context.Background(), 15*time.Second)
	defer cancel()

	log.Printf("Fetching HTML from %s...", targetURL)
	htmlContent, err := fetchHTML(ctx, targetURL)
	if err != nil {
		log.Fatalf("Failed to fetch HTML: %v", err)
	}

	// Simple regex for extracting titles (using title tags or similar, depending on the site)
	// This is a naive regex for demonstration on real sites
	re := regexp.MustCompile(`class="titleline"><a href=".*?">(.*?)</a>`)
	matches := re.FindAllStringSubmatch(htmlContent, -1)

	if len(matches) == 0 {
		log.Println("No titles found. Adjust regex for the target site.")
		return
	}

	log.Printf("Extracted %d titles. Saving to database...\n", len(matches))
	stmt, err := db.PrepareContext(ctx, "INSERT INTO article_titles(title) VALUES($1)")
	if err != nil {
		log.Fatalf("Failed to prepare statement: %v", err)
	}
	defer stmt.Close()

	for _, match := range matches {
		if len(match) > 1 {
			if _, err := stmt.ExecContext(ctx, match[1]); err != nil {
				log.Printf("Failed to insert title '%s': %v", match[1], err)
			}
		}
	}
	log.Println("Titles saved successfully.")
}
